modules/programs/carapace.nix

```diff
@@ -0,0 +1,62 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let cfg = config.programs.carapace;
+in {
+  meta.maintainers = [ maintainers.bobvanderlinden ];
+
+  options = {
+    programs.carapace = {
+      enable = mkEnableOption "carapace shell completion";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.carapace;
+        defaultText = literalExpression "pkgs.carapace";
+        description = ''
+          carapace package to use
+        '';
+      };
+
+      enableBashIntegration = mkOption {
+        default = true;
+        type = types.bool;
+        description = ''
+          Whether to enable Bash integration.
+        '';
+      };
+
+      enableZshIntegration = mkOption {
+        default = true;
+        type = types.bool;
+        description = ''
+          Whether to enable Zsh integration.
+        '';
+      };
+
+      enableFishIntegration = mkOption {
+        default = true;
+        type = types.bool;
+        description = ''
+          Whether to enable Fish integration.
+        '';
+      };
+    };
+  };
+  config = mkIf cfg.enable {
+    home.packages = [ cfg.package ];
+
+    programs.bash.initExtra = mkIf cfg.enableBashIntegration ''
+      source <(${cfg.package}/bin/carapace _carapace)
+    '';
+
+    programs.zsh.initExtra = mkIf cfg.enableZshIntegration ''
+      eval (${cfg.package}/bin/carapace _carapace|slurp)
+    '';
+
+    programs.fish.shellInit = mkIf cfg.enableFishIntegration ''
+      mkdir -p ~/.config/fish/completions
+      ${cfg.package}/bin/carapace --list | awk '{print $1}' | xargs -I{} touch ~/.config/fish/completions/{}.fish # disable auto-loaded completions (#185)
```

Comment:
This should probably be done in an activation script or by using import-from-derivation (`home.file.something.source = pkgs.runCommand ...`).

Reply:
Hmm I have now implemented this, but it now conflicts with  `.config/fish/completions` as a whole, instead of individual files.

Since this mechanism is merely to disable builtin autocompletion from fish itself and use carapace instead, it doesn't seem that nice.

I have rewritten it a bit so it now creates individual home.file entries based on the output of `carapace --list`.